Adding steps to follow when handling API/ABI breakage of a package (#1688)
conda-forge/conda-forge.github.io: src/user/faq.rst

  
  If you need to compile CUDA code, even if it involves only CUDA host APIs, you will still need a valid CUDA Toolkit installed locally and use it. Please refer to `NVCC's documentation <https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html>`_ for the CUDA compiler usage and `CUDA Programming Guide <https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html>`_ for general CUDA programming.

.. _faq_abi_incompatibility:

:ref:`(Q) <faq_abi_incompatibility>` **How to handle breaking of a package due to ABI incompatibility?**

  If your package breaks ABI with a version bump, here are a few steps you can take to fix it:

>   - Rebuild the package with corrected ``run_exports``.

Reviewer: ```suggestion
  - Rebuild the new version with corrected ``run_exports``.
  - Rebuild the old version with corrected ``run_exports``.
```

  Unfortunately, this is not possible with conda-forge's current infrastructure (``nvcc``, ``cudatoolkit``, etc) if there is no local CUDA Toolkit installation. In particular, the ``nvcc`` package provided on conda-forge is a *wrapper package* that exposes the actual ``nvcc`` compiler to our CI infrastructure in a ``conda``-friendly way; it does not contain the full ``nvcc`` compiler toolchain. One of the reasons is that CUDA headers like ``cuda.h``, ``cuda_runtime.h``, etc, which are needed at compile time, are not redistributable according to NVIDIA's EULA. Likewise, the ``cudatoolkit`` package only contains CUDA runtime libraries and not the compiler toolchain.
  
  If you need to compile CUDA code, even if it involves only CUDA host APIs, you will still need a valid CUDA Toolkit installed locally and use it. Please refer to `NVCC's documentation <https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html>`_ for the CUDA compiler usage and `CUDA Programming Guide <https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html>`_ for general CUDA programming.

.. _faq_api_abi_breakage:

> :ref:`(Q) <faq_api_abi_breakage>` **How to handle a ABI/API breakage of a package?**

Author: I am really sorry @isuruf, I can't seem to understand your question. Are you suggesting that we should also add in brief what could be the possible reasons for API/ABI breakage of a package? I am new here so, I am not very well versed with everything and a little confused as well, but I am learning, so I would be grateful if you could please help.

> :ref:`(Q) <faq_api_abi_breakage>` **How to handle a ABI/API breakage of a package?**

Reviewer: I mean is it a breakage in a version number (if so, is it covered by `run_exports`) or is it a breakage in a build number?

> :ref:`(Q) <faq_api_abi_breakage>` **How to handle a ABI/API breakage of a package?**

Reviewer 2: IIUC Surbhi is one of the Outreachy applicants so may not know relevant technical details (though I could be wrong about this)

If there is particular verbiage or structuring of this, would suggest we just propose those changes

> :ref:`(Q) <faq_api_abi_breakage>` **How to handle a ABI/API breakage of a package?**

Author: @jakirkham, that's right, I am an outreachy applicant. I am still learning and trying to understand how everything works, and hope to be able to help more in future, but till then I would be glad if you folks could help me with things.

> :ref:`(Q) <faq_api_abi_breakage>` **How to handle a ABI/API breakage of a package?**

Reviewer 3: This faq is supposed to cover what happens when we ship a package with a run export at say x.x, but we should have had that be say x.x.x. Typically this happens when there is an ABI break during a patch or minor version update. I think the corrected title below is better. 

We could also simply say how to fix incorrect run exports or something along those lines.

> :ref:`(Q) <faq_api_abi_breakage>` **How to handle a ABI/API breakage of a package?**

Author: Ah, okay! Thanks a lot, @beckermr, for explaining this. So, should I go with the same title for the FAQ or change it to "How to fix `run_exports` to handle ABI breakage?" as also suggested by Matt, @jakirkham @isuruf, what do you folks say?

> :ref:`(Q) <faq_api_abi_breakage>` **How to handle a ABI/API breakage of a package?**

Reviewer 3: Let's go with the title as is for now.

> :ref:`(Q) <faq_api_abi_breakage>` **How to handle breaking of a package due to ABI/API incompatibility?**

Reviewer: ```suggestion
:ref:`(Q) <faq_api_abi_breakage>` **How to handle breaking of a package due to ABI incompatibility?**
```

>   If your package breaks due to ABI/API incompatibility, here are a few steps you can take to fix it:

Reviewer: ```suggestion
  If your package breaks ABI with a version bump, here are a few steps you can take to fix it:
```

  - Rebuild the package with corrected ``run_exports``.

>   - Hot-fix repodata of previous package to apply the right ``run_exports``.

Reviewer: `run_exports` cannot be hot-fixed

>   - Hot-fix the repodata of dependencies to include corrected pinnings for the package.

Reviewer 2: ```suggestion
  - Hot-fix the repodata of dependencies to include corrected pinnings for the package.
  - Add a PR to pin the old version in ``conda-forge-pinning`` (if not already present)
  - Open a migrator following `CFEP-09 <https://github.com/conda-forge/cfep/blob/main/cfep-09.md>`_
```
